File: crypto/file_tags.go
package crypto

import (
	_ "embed"
	"encoding/json"
	"fmt"
	"regexp"
	"strings"
	"sync"
)

//go:embed file-tags-params.json
var embeddedFileTagsParams []byte

// FileTagsParams holds shared tag limits for Go and TypeScript clients.
type FileTagsParams struct {
	MaxTagsPerFile        int `json:"maxTagsPerFile"`
	MaxTagLength          int `json:"maxTagLength"`
	MaxTagsPerFilterQuery int `json:"maxTagsPerFilterQuery"`
}

// MaxEncryptedTagsBase64Len is the server-side cap on non-empty encrypted_tags
// base64 strings (abuse guard; plaintext limits are enforced only by clients).
const MaxEncryptedTagsBase64Len = 1024

// AES-GCM nonce size for metadata fields (12 bytes -> 16 base64 characters).
const TagsNonceRawBytes = 12

var (
	fileTagsParamsOnce  sync.Once
	fileTagsParamsCache *FileTagsParams
	fileTagsParamsErr   error

	tagSyntaxRE = regexp.MustCompile(`^[A-Za-z0-9]+(?:-[A-Za-z0-9]+)*$`)
)

// LoadFileTagsParams loads tag limits from the embedded JSON.
func LoadFileTagsParams() (*FileTagsParams, error) {
	fileTagsParamsOnce.Do(func() {
		params := &FileTagsParams{}
		if err := json.Unmarshal(embeddedFileTagsParams, params); err != nil {
			fileTagsParamsErr = fmt.Errorf("failed to parse embedded file-tags-params.json: %w", err)
			return
		}
		if params.MaxTagsPerFile <= 0 || params.MaxTagLength <= 0 || params.MaxTagsPerFilterQuery <= 0 {
			fileTagsParamsErr = fmt.Errorf("file-tags-params.json has invalid non-positive limits")
			return
		}
		fileTagsParamsCache = params
	})
	return fileTagsParamsCache, fileTagsParamsErr
}

// GetFileTagsParams returns loaded tag limits or panics.
func GetFileTagsParams() *FileTagsParams {
	params, err := LoadFileTagsParams()
	if err != nil {
		panic(fmt.Sprintf("Failed to load file tags params: %v", err))
	}
	return params
}

// GetEmbeddedFileTagsParamsJSON returns the raw embedded JSON for API serving.
func GetEmbeddedFileTagsParamsJSON() []byte {
	return embeddedFileTagsParams
}

// ParseTagList splits a comma-separated tag string, trims surrounding whitespace
// per segment, and rejects empty segments. It does not validate syntax or limits.
func ParseTagList(input string) ([]string, error) {
	if strings.TrimSpace(input) == "" {
		return []string{}, nil
	}
	parts := strings.Split(input, ",")
	out := make([]string, 0, len(parts))
	for _, part := range parts {
		tag := strings.TrimSpace(part)
		if tag == "" {
			return nil, fmt.Errorf("empty tag segment")
		}
		out = append(out, tag)
	}
	return out, nil
}

// ValidateTagSyntax checks a single tag against the locked character rules and length.
// Failures name the specific rule that was violated.
func ValidateTagSyntax(tag string, maxLen int) error {
	if tag == "" {
		return fmt.Errorf("tag is empty")
	}
	if maxLen <= 0 {
		maxLen = GetFileTagsParams().MaxTagLength
	}
	if len(tag) > maxLen {
		return fmt.Errorf("tag exceeds max length %d", maxLen)
	}
	if strings.ContainsAny(tag, " \t\n\r") {
		return fmt.Errorf("tag contains whitespace")
	}
	if strings.HasPrefix(tag, "-") {
		return fmt.Errorf("tag cannot start with a dash")
	}
	if strings.HasSuffix(tag, "-") {
		return fmt.Errorf("tag cannot end with a dash")
	}
	if strings.Contains(tag, "--") {
		return fmt.Errorf("tag cannot contain consecutive dashes")
	}
	for _, r := range tag {
		if (r >= 'A' && r <= 'Z') || (r >= 'a' && r <= 'z') || (r >= '0' && r <= '9') || r == '-' {
			continue
		}
		return fmt.Errorf("tag contains invalid characters (use A-Z, a-z, 0-9, and single dashes between segments)")
	}
	if !tagSyntaxRE.MatchString(tag) {
		return fmt.Errorf("tag has invalid syntax (use alphanumeric segments separated by single dashes)")
	}
	return nil
}

// ErrMaxTagsPerFile returns the canonical user-facing limit error.
func ErrMaxTagsPerFile(max int) error {
	return fmt.Errorf("%d tags maximum", max)
}

// CanonicalizeTags validates, deduplicates case-insensitively (first-seen casing),
// enforces max tags per file, and returns the canonical ordered list.
func CanonicalizeTags(tags []string) ([]string, error) {
	params := GetFileTagsParams()
	if len(tags) > params.MaxTagsPerFile {
		return nil, ErrMaxTagsPerFile(params.MaxTagsPerFile)
	}
	seen := make(map[string]struct{}, len(tags))
	out := make([]string, 0, len(tags))
	for _, tag := range tags {
		if err := ValidateTagSyntax(tag, params.MaxTagLength); err != nil {
			return nil, err
		}
		key := strings.ToLower(tag)
		if _, ok := seen[key]; ok {
			continue
		}
		seen[key] = struct{}{}
		out = append(out, tag)
	}
	if len(out) > params.MaxTagsPerFile {
		return nil, ErrMaxTagsPerFile(params.MaxTagsPerFile)
	}
	return out, nil
}

// ParseAndCanonicalizeTags parses input then canonicalizes.
func ParseAndCanonicalizeTags(input string) ([]string, error) {
	parsed, err := ParseTagList(input)
	if err != nil {
		return nil, err
	}
	return CanonicalizeTags(parsed)
}

// SerializeTags joins tags with commas and no surrounding spaces.
func SerializeTags(tags []string) string {
	return strings.Join(tags, ",")
}

// ParseFilterTags parses a filter query, collapses case-insensitive duplicates
// (first-seen casing), and enforces the filter-query max against the unique set.
func ParseFilterTags(input string) ([]string, error) {
	params := GetFileTagsParams()
	parsed, err := ParseTagList(input)
	if err != nil {
		return nil, err
	}
	seen := make(map[string]struct{}, len(parsed))
	out := make([]string, 0, len(parsed))
	for _, tag := range parsed {
		if err := ValidateTagSyntax(tag, params.MaxTagLength); err != nil {
			return nil, err
		}
		key := strings.ToLower(tag)
		if _, ok := seen[key]; ok {
			continue
		}
		seen[key] = struct{}{}
		out = append(out, tag)
	}
	if len(out) > params.MaxTagsPerFilterQuery {
		return nil, fmt.Errorf("too many filter tags: max %d", params.MaxTagsPerFilterQuery)
	}
	return out, nil
}

// FileHasAllTags reports whether fileTags contains every query tag (AND),
// using ASCII case-folded comparison.
func FileHasAllTags(fileTags, queryTags []string) bool {
	if len(queryTags) == 0 {
		return true
	}
	set := make(map[string]struct{}, len(fileTags))
	for _, t := range fileTags {
		set[strings.ToLower(t)] = struct{}{}
	}
	for _, q := range queryTags {
		if _, ok := set[strings.ToLower(q)]; !ok {
			return false
		}
	}
	return true
}

// AddTag appends tag if not already present (case-insensitive). Preserves order.
func AddTag(tags []string, tag string) ([]string, error) {
	return AddTags(tags, []string{tag})
}

// AddTags appends each tag in toAdd if not already present (case-insensitive).
// Preserves existing order and appends new tags in toAdd order. Whitespace inside
// a tag remains invalid; callers should ParseTagList first so spaces around commas
// are trimmed. Returns ErrMaxTagsPerFile if the resulting list would exceed the limit.
func AddTags(tags []string, toAdd []string) ([]string, error) {
	params := GetFileTagsParams()
	out := append([]string(nil), tags...)
	for _, tag := range toAdd {
		if err := ValidateTagSyntax(tag, params.MaxTagLength); err != nil {
			return nil, err
		}
		key := strings.ToLower(tag)
		present := false
		for _, t := range out {
			if strings.ToLower(t) == key {
				present = true
				break
			}
		}
		if present {
			continue
		}
		if len(out) >= params.MaxTagsPerFile {
			return nil, ErrMaxTagsPerFile(params.MaxTagsPerFile)
		}
		out = append(out, tag)
	}
	return out, nil
}

// ParseAndAddTags parses a comma-separated input (trimming spaces around commas)
// and merges the tags into existing via AddTags.
func ParseAndAddTags(existing []string, input string) ([]string, error) {
	parsed, err := ParseTagList(input)
	if err != nil {
		return nil, err
	}
	if len(parsed) == 0 {
		return nil, fmt.Errorf("tag is empty")
	}
	return AddTags(existing, parsed)
}

// RemoveTag removes the first case-insensitive match. Missing tag is a no-op.
func RemoveTag(tags []string, tag string) []string {
	key := strings.ToLower(tag)
	out := make([]string, 0, len(tags))
	removed := false
	for _, t := range tags {
		if !removed && strings.ToLower(t) == key {
			removed = true
			continue
		}
		out = append(out, t)
	}
	return out
}

// ReplaceTag replaces the first case-insensitive match for oldTag with newTag,
// keeping position. Casing-only replacement is allowed. Collision with another
// existing tag (different from oldTag) is rejected.
func ReplaceTag(tags []string, oldTag, newTag string) ([]string, error) {
	params := GetFileTagsParams()
	if err := ValidateTagSyntax(newTag, params.MaxTagLength); err != nil {
		return nil, err
	}
	oldKey := strings.ToLower(oldTag)
	newKey := strings.ToLower(newTag)
	idx := -1
	for i, t := range tags {
		if strings.ToLower(t) == oldKey {
			idx = i
			break
		}
	}
	if idx < 0 {
		return nil, fmt.Errorf("tag not found")
	}
	if oldKey != newKey {
		for i, t := range tags {
			if i != idx && strings.ToLower(t) == newKey {
				return nil, fmt.Errorf("replacement collides with existing tag")
			}
		}
	}
	out := make([]string, len(tags))
	copy(out, tags)
	out[idx] = newTag
	return out, nil
}

// TagPresent reports case-insensitive membership.
func TagPresent(tags []string, tag string) bool {
	key := strings.ToLower(tag)
	for _, t := range tags {
		if strings.ToLower(t) == key {
			return true
		}
	}
	return false
}

File: client/static/js/src/crypto/file-tags.ts
/**
 * Owner file tag parse / validate / mutate helpers.
 * Limits come from crypto/file-tags-params.json via GET /api/config/file-tags.
 * No hardcoded browser fallback: callers must handle load failure.
 */

export interface FileTagsParams {
  maxTagsPerFile: number;
  maxTagLength: number;
  maxTagsPerFilterQuery: number;
}

const TAG_SYNTAX = /^[A-Za-z0-9]+(?:-[A-Za-z0-9]+)*$/;

let cachedParams: FileTagsParams | null = null;
let loadPromise: Promise<FileTagsParams> | null = null;

export function clearFileTagsParamsCacheForTests(): void {
  cachedParams = null;
  loadPromise = null;
}

export async function loadFileTagsParams(): Promise<FileTagsParams> {
  if (cachedParams) {
    return cachedParams;
  }
  if (!loadPromise) {
    loadPromise = (async () => {
      const response = await fetch('/api/config/file-tags');
      if (!response.ok) {
        throw new Error(`Failed to load file-tags config: HTTP ${response.status}`);
      }
      const data = (await response.json()) as FileTagsParams;
      if (
        !data ||
        typeof data.maxTagsPerFile !== 'number' ||
        typeof data.maxTagLength !== 'number' ||
        typeof data.maxTagsPerFilterQuery !== 'number' ||
        data.maxTagsPerFile <= 0 ||
        data.maxTagLength <= 0 ||
        data.maxTagsPerFilterQuery <= 0
      ) {
        throw new Error('Invalid file-tags config payload');
      }
      cachedParams = data;
      return data;
    })().catch((err) => {
      loadPromise = null;
      throw err;
    });
  }
  return loadPromise;
}

export function getCachedFileTagsParams(): FileTagsParams | null {
  return cachedParams;
}

export function parseTagList(input: string): string[] {
  if (input.trim() === '') {
    return [];
  }
  const parts = input.split(',');
  const out: string[] = [];
  for (const part of parts) {
    const tag = part.trim();
    if (tag === '') {
      throw new Error('empty tag segment');
    }
    out.push(tag);
  }
  return out;
}

export function validateTagSyntax(tag: string, maxLen: number): void {
  if (!tag) {
    throw new Error('tag is empty');
  }
  if (tag.length > maxLen) {
    throw new Error(`tag exceeds max length ${maxLen}`);
  }
  if (/\s/.test(tag)) {
    throw new Error('tag contains whitespace');
  }
  if (!TAG_SYNTAX.test(tag)) {
    throw new Error('tag has invalid syntax');
  }
}

/** Canonical user-facing limit error for per-file tag count. */
export function maxTagsPerFileError(max: number): Error {
  return new Error(`${max} tags maximum`);
}

export function canonicalizeTags(tags: string[], params: FileTagsParams): string[] {
  if (tags.length > params.maxTagsPerFile) {
    throw maxTagsPerFileError(params.maxTagsPerFile);
  }
  const seen = new Set<string>();
  const out: string[] = [];
  for (const tag of tags) {
    validateTagSyntax(tag, params.maxTagLength);
    const key = tag.toLowerCase();
    if (seen.has(key)) {
      continue;
    }
    seen.add(key);
    out.push(tag);
  }
  if (out.length > params.maxTagsPerFile) {
    throw maxTagsPerFileError(params.maxTagsPerFile);
  }
  return out;
}

export function parseAndCanonicalizeTags(input: string, params: FileTagsParams): string[] {
  return canonicalizeTags(parseTagList(input), params);
}

export function serializeTags(tags: string[]): string {
  return tags.join(',');
}

export function parseFilterTags(input: string, params: FileTagsParams): string[] {
  const parsed = parseTagList(input);
  const seen = new Set<string>();
  const out: string[] = [];
  for (const tag of parsed) {
    validateTagSyntax(tag, params.maxTagLength);
    const key = tag.toLowerCase();
    if (seen.has(key)) {
      continue;
    }
    seen.add(key);
    out.push(tag);
  }
  if (out.length > params.maxTagsPerFilterQuery) {
    throw new Error(`too many filter tags: max ${params.maxTagsPerFilterQuery}`);
  }
  return out;
}

export function fileHasAllTags(fileTags: string[], queryTags: string[]): boolean {
  if (queryTags.length === 0) {
    return true;
  }
  const set = new Set(fileTags.map((t) => t.toLowerCase()));
  return queryTags.every((q) => set.has(q.toLowerCase()));
}

export function tagPresent(tags: string[], tag: string): boolean {
  const key = tag.toLowerCase();
  return tags.some((t) => t.toLowerCase() === key);
}

export function addTag(tags: string[], tag: string, params: FileTagsParams): string[] {
  return addTags(tags, [tag], params);
}

/**
 * Append each tag in toAdd if not already present (case-insensitive).
 * Preserves existing order; appends new tags in toAdd order.
 * Callers should parseTagList first so spaces around commas are trimmed.
 */
export function addTags(tags: string[], toAdd: string[], params: FileTagsParams): string[] {
  const out = tags.slice();
  for (const tag of toAdd) {
    validateTagSyntax(tag, params.maxTagLength);
    if (tagPresent(out, tag)) {
      continue;
    }
    if (out.length >= params.maxTagsPerFile) {
      throw maxTagsPerFileError(params.maxTagsPerFile);
    }
    out.push(tag);
  }
  return out;
}

/** Parse comma-separated input (trim around commas) and merge into existing tags. */
export function parseAndAddTags(existing: string[], input: string, params: FileTagsParams): string[] {
  const parsed = parseTagList(input);
  if (parsed.length === 0) {
    throw new Error('tag is empty');
  }
  return addTags(existing, parsed, params);
}

export function removeTag(tags: string[], tag: string): string[] {
  const key = tag.toLowerCase();
  let removed = false;
  return tags.filter((t) => {
    if (!removed && t.toLowerCase() === key) {
      removed = true;
      return false;
    }
    return true;
  });
}

export function replaceTag(
  tags: string[],
  oldTag: string,
  newTag: string,
  params: FileTagsParams,
): string[] {
  validateTagSyntax(newTag, params.maxTagLength);
  const oldKey = oldTag.toLowerCase();
  const newKey = newTag.toLowerCase();
  const idx = tags.findIndex((t) => t.toLowerCase() === oldKey);
  if (idx < 0) {
    throw new Error('tag not found');
  }
  if (oldKey !== newKey) {
    for (let i = 0; i < tags.length; i++) {
      if (i !== idx && tags[i].toLowerCase() === newKey) {
        throw new Error('replacement collides with existing tag');
      }
    }
  }
  const out = tags.slice();
  out[idx] = newTag;
  return out;
}

export interface TagVocabularyEntry {
  display: string;
  count: number;
}

/** Build case-insensitive vocabulary with display spelling and usage counts. */
export function buildTagVocabulary(fileTagLists: Array<string[] | null | undefined>): TagVocabularyEntry[] {
  const map = new Map<string, TagVocabularyEntry>();
  for (const list of fileTagLists) {
    if (!list) continue;
    for (const tag of list) {
      const key = tag.toLowerCase();
      const existing = map.get(key);
      if (existing) {
        existing.count += 1;
      } else {
        map.set(key, { display: tag, count: 1 });
      }
    }
  }
  return Array.from(map.values());
}

const DEFAULT_SUGGESTION_LIMIT = 8;

export function suggestTags(
  vocabulary: TagVocabularyEntry[],
  query: string,
  selected: string[],
  limit: number = DEFAULT_SUGGESTION_LIMIT,
): string[] {
  const q = query.trim().toLowerCase();
  const selectedSet = new Set(selected.map((t) => t.toLowerCase()));
  const candidates = vocabulary.filter((v) => !selectedSet.has(v.display.toLowerCase()));
  const scored = candidates
    .map((v) => {
      const displayLower = v.display.toLowerCase();
      let rank = 2;
      if (q && displayLower.startsWith(q)) {
        rank = 0;
      } else if (q && displayLower.includes(q)) {
        rank = 1;
      } else if (q) {
        return null;
      }
      return { display: v.display, rank, count: v.count };
    })
    .filter((x): x is { display: string; rank: number; count: number } => x !== null);
  scored.sort((a, b) => {
    if (a.rank !== b.rank) return a.rank - b.rank;
    if (b.count !== a.count) return b.count - a.count;
    return a.display.localeCompare(b.display);
  });
  return scored.slice(0, limit).map((s) => s.display);
}
